lib/testing/matrix.js

```diff
@@ -0,0 +1,172 @@
+        this.callback = callback;
+        return this;
+    }
+
+    /**
+    * Method to prevent an exception during the tests.
+    *
+    * @param {Object.<string, string[]>} exception - The keys of the hash table
+    is the element we want to do an exception, the value
+    *                      is the value we want to do an exception for the
+    *                      specific key.
+    *                      All key of the object must be validate with at least
+    *                      one value for execute an exception.
+    *                      ex: for a hashmap with 'auth' as key and 'v4' as
+    *                      value, if we found a matrix with this following
+    *                      paramater, it will call the special callback
+    * @argument {GenerateMatrixCallback} callbackException - Callback who the
+    *                       execute method will call when
+    *                       this exception is found.
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    if(exception, callbackException) {
+        const pair = {
+            key: exception,
+            callback: callbackException,
+        };
+        this.listOfExceptions.push(pair);
+
+        return this;
+    }
+
+    /**
+    * Return the parameters of the matrix into a Json format
+    *
+    * @returns {string} JSON format of parameters
+    */
+    serialize() {
+        return this.paramsToString;
+    }
+
+    /**
+    * Generate the matrix and call all callback enter on @generate and
+    * if methods. Must be run as the father matrix for running all the test
+    *
+    * @returns {undefined}
+    */
+    execute() {
+        /**
+        * Check if @generate method has been already called
+        */
+        if (this.callback) {
+            /**
+            * Generate matrix element by element, we call the @execute
+            * method until that we specialize all element send
+            * by generate method
+            */
+            if (this.elementsToSpecialize.length !== 0) {
+                /**
+                * Make copy for restore object in the end of recursive function
+                */
+                const elementToSpecialize = this.elementsToSpecialize[0];
+                const copyElement = this.params[elementToSpecialize];
+
+                this.elementsToSpecialize.shift();
+                if (!(elementToSpecialize in this.params)) {
+                    const errMessage = 'No key of specialization found: ';
+                    throw new Error(errMessage + elementToSpecialize);
+                } else {
+                    copyElement.forEach(currentElement => {
+                        this.params[elementToSpecialize] = currentElement;
+                        this.execute();
+                    });
+                }
+                this.params[elementToSpecialize] = copyElement;
+                this.elementsToSpecialize.push(elementToSpecialize);
```

Comment:
Aren't we changing the order of the specialized parameters here ? (I'm not sure it's necessary relevant, but we should try to avoid altering things as much as possible)
First, we extract the index 0 (first), then we push it back at the last index.

Another method, rather than shift() and push() would be to modify a copy of the original array, and restore the reference to the original array at the end:
```
const origSpecList = this.elementsToSpecialize;
this.elementsToSpecialize = origSpecList.slice(0);

[do stuff that modifies this.elementsToSpecialize...]

this.elementsToSpecialize = origSpecList;
```